nab/detectors/htmcore/htmcore_detector.py

```diff
@@ -0,0 +1,135 @@
+
+import math
+
+# htm.core imports
+from htm.bindings.algorithms import SpatialPooler   as SP
+from htm.bindings.algorithms import TemporalMemory  as TM
+from htm.bindings.algorithms import Predictor
+
+from htm.algorithms.anomaly_likelihood import AnomalyLikelihood
+from nab.detectors.base import AnomalyDetector
+
+# Fraction outside of the range of values seen so far that will be considered
+# a spatial anomaly regardless of the anomaly likelihood calculation. This
+# accounts for the human labelling bias for spatial values larger than what
+# has been seen so far.
+SPATIAL_TOLERANCE = 0.05
+
+
+
+class HtmcoreDetector(AnomalyDetector):
+  """
+  This detector uses an HTM based anomaly detection technique.
+  """
+
+  def __init__(self, *args, **kwargs):
+
+    super(HtmcoreDetector, self).__init__(*args, **kwargs)
+
+    self.anomalyLikelihood = None
+    # Keep track of value range for spatial anomaly detection
+    self.minVal = None
+    self.maxVal = None
+
+    # Set this to False if you want to get results based on raw scores
+    # without using AnomalyLikelihood. This will give worse results, but
+    # useful for checking the efficacy of AnomalyLikelihood. You will need
+    # to re-optimize the thresholds when running with this setting.
+    self.useLikelihood = True
+
+
+  def getAdditionalHeaders(self):
+    """Returns a list of strings."""
+    return ["raw_score"] #TODO add "prediction"
+
+
+  def handleRecord(self, inputData):
+    """Returns a tuple (anomalyScore, rawScore).
+
+    Internally to NuPIC "anomalyScore" corresponds to "likelihood_score"
+    and "rawScore" corresponds to "anomaly_score". Sorry about that.
+    """
+    # Send it to Numenta detector and get back the results
+    result = [] #FIXME self.model.run(inputData)
+
+    # Get the value
+    value = inputData["value"]
+
+    # Retrieve the anomaly score and write it to a file
+    rawScore = 0.5 #FIXME result.inferences["anomalyScore"]
+
+    # Update min/max values and check if there is a spatial anomaly
+    spatialAnomaly = False #TODO make this computed in SP (and later improve)
+    if self.minVal != self.maxVal:
+      tolerance = (self.maxVal - self.minVal) * SPATIAL_TOLERANCE
+      maxExpected = self.maxVal + tolerance
+      minExpected = self.minVal - tolerance
+      if value > maxExpected or value < minExpected:
+        spatialAnomaly = True
+    if self.maxVal is None or value > self.maxVal:
+      self.maxVal = value
+    if self.minVal is None or value < self.minVal:
+      self.minVal = value
+
+    if self.useLikelihood:
+      # Compute log(anomaly likelihood)
+      anomalyScore = self.anomalyLikelihood.anomalyProbability(inputData["value"], rawScore, inputData["timestamp"])
+      logScore = self.anomalyLikelihood.computeLogLikelihood(anomalyScore)
+      finalScore = logScore #TODO returls logScore and not probability? Fix that in our Likelihood.cpp; #TODO TM to provide anomaly {none, raw, likelihood} 
```

Comment:
- [ ] we should probably return logScore in Likelihood by default https://github.com/htm-community/htm.core/issues/469
- [ ] provide enum modes {none, raw, likelihood} to `TM.anomaly`

Reply:
both done in https://github.com/htm-community/htm.core/pull/588